cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go

```diff
@@ -226,7 +234,64 @@ func (ng *nodegroup) Nodes() ([]cloudprovider.Instance, error) {
 // node by default, using manifest (most likely only kube-proxy).
 // Implementation optional.
 func (ng *nodegroup) TemplateNodeInfo() (*schedulerframework.NodeInfo, error) {
-	return nil, cloudprovider.ErrNotImplemented
+	if !ng.scalableResource.CanScaleFromZero() {
+		return nil, cloudprovider.ErrNotImplemented
+	}
+
+	capacity, err := ng.scalableResource.InstanceCapacity()
+	if err != nil {
+		return nil, err
+	}
+
+	_, cpuPresent := capacity[corev1.ResourceCPU]
+	_, memPresent := capacity[corev1.ResourceMemory]
+	if !cpuPresent || !memPresent {
+		return nil, cloudprovider.ErrNotImplemented
+	}
+
+	nodeName := fmt.Sprintf("%s-asg-%d", ng.scalableResource.Name(), rand.Int63())
+	node := corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   nodeName,
+			Labels: map[string]string{},
+		},
+	}
+
+	node.Status.Capacity = capacity
+	node.Status.Allocatable = capacity
+	node.Status.Conditions = cloudprovider.BuildReadyConditions()
+	node.Spec.Taints = ng.scalableResource.Taints()
+
+	node.Labels, err = ng.buildTemplateLabels(nodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeInfo := schedulerframework.NewNodeInfo(cloudprovider.BuildKubeProxy(ng.scalableResource.Name()))
+	nodeInfo.SetNode(&node)
+
+	return nodeInfo, nil
+}
+
+func (ng *nodegroup) buildTemplateLabels(nodeName string) (map[string]string, error) {
+	labels := cloudprovider.JoinStringMaps(ng.scalableResource.Labels(), buildGenericLabels(nodeName))
+
+	nodes, err := ng.Nodes()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(nodes) > 0 {
+		node, err := ng.machineController.findNodeByProviderID(normalizedProviderString(nodes[0].Id))
+		if err != nil {
+			return nil, err
+		}
+
+		if node != nil {
+			labels = cloudprovider.JoinStringMaps(labels, extractNodeLabels(node))
+		}
+	}
+	return labels, nil
```

Comment:
reading through the various labels funcs here it seems like there will never be a situation where the labels `map[string]string` will be nil (at present `buildGenericLabels` will always guarantee a bunch of labels), but do we want to be a bit more defensive about that and return an error if for any reason our `labels` var ends up nil at the end of this function?

Reply:
i need to think that through a little bit more, but it sounds like a good defensive practice. the idea would be that if we don't see the labels we expect, then we would return some sort of error (which would cause the scaling to not happen) ?

Reply:
It's certainly valid for the `nodeInfo` object (type `*schedulerframework.NodeInfo
`) to not have labels.

In any event this is probably academic as the UT in `cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go` will protect us from this scenario. I think we're fine.

Reply:
> It's certainly valid for the nodeInfo object (type *schedulerframework.NodeInfo ) to not have labels.

agreed, my concern is messing up the simulator logic that uses the labels for matching. i'll do a little more investigation.